Move competition events to a dedicated page - Fixes #937

pingskills opened the pull request:
There are plans to store entry fees against each competition event
and this will make that easier.

pingskills commented:
I wasn't sure the best way to go about this so I'd appreciate some feedback. Should I be creating new routes like this? Should I be trying to use the existing update method?

Here's the new page and the new menu item on the left
![edit-events](https://cloud.githubusercontent.com/assets/4403168/19303981/88956f8e-90b6-11e6-9795-3064828df04a.png)

If you aren't an admin and the competition is confirmed you get the following message:
![edit-events-organiser-after-confirmed](https://cloud.githubusercontent.com/assets/4403168/19304007/a580c026-90b6-11e6-8a91-5e6026d032d2.png)

jonatanklosko commented:
I think it would be good to show an alert when the event list is empty. It could contain a link to the events page, so organizers/delegates know what's going on.

jfly commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -234,6 +234,20 @@ def post_results
+    @competition = Competition.find(params[:id])
+  end
+
+  def save_events
```
Nit: For consistency, I think I'd use the verb "update" here.

jfly commented on WcaOnRails/app/views/competitions/_competition_form.html.erb:
```diff
@@ -276,7 +255,7 @@
         <% end %>
 
-        <% if @competition.persisted? && !@competition_admin_view && !is_actually_confirmed && current_user.can_confirm_competition?(@competition) %>
+        <% if @competition.persisted? && !@competition_admin_view && !is_actually_confirmed && current_user.can_confirm_competition?(@competition) && @competition.events.count > 0 %>
```
Hmmm... this feels like a duplication of this logic we already have in our Competition model: https://github.com/thewca/worldcubeassociation.org/blob/8b7e8e363606b576735aa73a90cf8815c8d1b22f/WcaOnRails/app/models/competition.rb#L99-L111.

I think it would be better to present the user with the button and when they fail to confirm, they get a good error message that makes it clear they need to add events. @jonatanklosko's suggestion to show an alert when there are currently no events might go well with this.

jfly commented on WcaOnRails/config/routes.rb:
```diff
@@ -51,6 +51,8 @@
 
   get 'competitions/:id/edit/admin' => 'competitions#admin_edit', as: :admin_edit_competition
+  get 'competitions/:id/edit/events' => 'competitions#edit_events', as: :edit_competition_events
+  patch 'competitions/:id/save/events' => 'competitions#save_events', as: :save_competition_events
```
I'm not a REST fanatic, but I think the order of these last 2 pieces of the url should be switched. Something like:

```ruby
get 'competitions/:id/events/edit' => 'competitions#edit_events', as: :edit_competition_events
patch 'competitions/:id/events' => 'competitions#update_events', as: :update_competition_events
```

I believe that's more consistent with how our other edit/update routes work.

pingskills commented:
I added a warning when a competition has no events, and added some tests.

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -98,11 +98,15 @@ class Competition < ActiveRecord::Base
   end
 
+  def no_events
+    return competition_events.reject(&:marked_for_destruction?).empty?
```
No need for an explicit `return` here.

jfly replied:
Also, the ruby convention for methods that return a boolean value is to always end their name with a question mark, so `no_events` -> `no_events?`.

jonatanklosko commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -660,4 +660,64 @@
+      it 'does not allow access' do
+        expect {
+          get :edit_events, id: competition.id
+        }.to raise_error(ActionController::RoutingError)
```
IMO we should redirect instead.

jfly replied:
I agree, I think that's more consistent with how we do stuff elsewhere. In fact, it looks like you're using `redirect_unless_user`, so I'm surprised we'd get a `RoutingError` here.

pingskills replied:
Hmmm, I'll take a look at what is going on.

pingskills replied:
`  before_action -> { redirect_unless_user(:can_manage_competition?, competition_from_params) }, only: [:edit, :update, :edit_events, :update_events]`

This calls `competition_from_params`

`  private def competition_from_params
    competition = Competition.find(params[:competition_id] || params[:id])
    if !competition.user_can_view?(current_user)
      raise ActionController::RoutingError.new('Not Found')
    end
    competition
  end`

which gives a routing error is the user can't view the competition. Do you think this is fine or should I change the way this is being processed?

jfly commented on WcaOnRails/spec/models/competition_spec.rb:
```diff
@@ -199,6 +199,11 @@
       expect(competition).to be_valid
       expect(competition.warnings_for(nil)[:invisible]).to eq "This competition is not visible to the public."
     end
+    it "warns if competition has no events" do
```
extreme nit: i think our `it`s usually have 1 empty line before them

pingskills commented:
Updated the code based on feedback and also changed what happens when a user can't edit the events. I now show the form with everything disabled so they can see what events are selected but not update them. Otherwise there was no place to see this information.

Here's a screenshot of the disabled state:
![disabled](https://cloud.githubusercontent.com/assets/4403168/19582225/95142fce-977e-11e6-8719-746640d8b124.png)

coder13 commented:
Should we supporting changing events after a comp has been confirmed? should it be up to a committee or the board to change events after a comp has been confirmed?

pingskills commented:
@coder13 The way the code is now, once a competition has been confirmed it can only be updated through the administrator view and not by the competition organisers. The image shows the disabled form and the message "You can't manage a competitions events after it has been confirmed".

jonatanklosko commented:
LGTM =) Let's see what @jfly thinks.
